File: winui/Engine.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace MThreadDraw;

/// <summary>
/// The MThread Draw engine, running as a child process and spoken to over a pipe.
/// </summary>
/// <remarks>
/// <para>Everything interesting - tracing an image, finding a device, injecting
/// touches - lives in the Python engine, which the other platforms use too.
/// This class is the whole of the Windows side's knowledge of it: send a JSON
/// line, wait for the reply with the same id, and raise an event for anything
/// that arrives unasked.</para>
///
/// <para>The alternative was porting the engine to C#, which would mean two
/// implementations of the same subtleties, and they would drift apart by the
/// second release.</para>
/// </remarks>
public sealed class Engine : IAsyncDisposable
{
    private readonly Process _process;
    private readonly Dictionary<int, TaskCompletionSource<JsonNode>> _pending = new();
    private readonly SemaphoreSlim _writeLock = new(1, 1);
    private int _nextId;

    public event Action<string>? StatusChanged;
    public event Action<int, int>? ProgressChanged;
    public event Action<string>? Failed;
    public event Action<string, int, int>? FrameReady;
    public event Action<string>? MirrorLost;

    private Engine(Process process)
    {
        _process = process;
        _ = Task.Run(ReadLoopAsync);
    }

    /// <summary>Where the engine is, in a release and in a source checkout.</summary>
    public static (string file, string arguments) Locate()
    {
        var here = AppContext.BaseDirectory;

        // A release ships the engine as one executable beside the app.
        var packaged = Path.Combine(here, "engine", "mthread-draw-engine.exe");
        if (File.Exists(packaged))
        {
            return (packaged, string.Empty);
        }

        // A source checkout runs it from the repository's virtual environment,
        // and failing that from whatever python is on PATH.
        var repository = FindRepository();
        var venv = repository is null ? null : Path.Combine(repository, "venv", "Scripts", "python.exe");
        var python = venv is not null && File.Exists(venv) ? venv : "python";
        return (python, "-m mthread_draw.server");
    }

    /// <summary>Walk up from the binary looking for the engine's own source.</summary>
    /// <remarks>
    /// Counting "..\..\.." from the output directory works until somebody
    /// changes the configuration, the framework or the runtime identifier, at
    /// which point it silently points at the wrong folder. A marker file does
    /// not have that problem.
    /// </remarks>
    public static string? FindRepository()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null)
        {
            if (File.Exists(Path.Combine(directory.FullName, "mthread_draw", "server.py")))
            {
                return directory.FullName;
            }
            directory = directory.Parent;
        }
        return null;
    }

    public static async Task<Engine> StartAsync()
    {
        var (file, arguments) = Locate();
        var repository = FindRepository();

        var info = new ProcessStartInfo(file, arguments)
        {
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            WorkingDirectory = repository ?? AppContext.BaseDirectory,
        };

        var process = Process.Start(info)
                      ?? throw new InvalidOperationException($"Could not start the engine: {file}");

        var engine = new Engine(process);
        // The engine announces itself; if that never arrives, nothing else will.
        var ready = await engine.WaitForReadyAsync().ConfigureAwait(false);
        if (!ready)
        {
            var complaint = await process.StandardError.ReadToEndAsync().ConfigureAwait(false);
            throw new InvalidOperationException(
                $"The engine did not start.\n{file} {arguments}\n{complaint}");
        }
        return engine;
    }

    private TaskCompletionSource<bool> _ready = new();

    private Task<bool> WaitForReadyAsync()
    {
        return Task.WhenAny(_ready.Task, Task.Delay(TimeSpan.FromSeconds(30)))
                   .ContinueWith(t => _ready.Task.IsCompletedSuccessfully && _ready.Task.Result);
    }

    private async Task ReadLoopAsync()
    {
        try
        {
            while (await _process.StandardOutput.ReadLineAsync().ConfigureAwait(false) is { } line)
            {
                if (line.Length == 0)
                {
                    continue;
                }

                JsonNode? message;
                try
                {
                    message = JsonNode.Parse(line);
                }
                catch (JsonException)
                {
                    continue;
                }
                if (message is null)
                {
                    continue;
                }

                if (message["event"] is { } name)
                {
                    Dispatch(name.GetValue<string>(), message);
                    continue;
                }

                var id = message["id"]?.GetValue<int>();
                if (id is not null)
                {
                    TaskCompletionSource<JsonNode>? waiting;
                    lock (_pending)
                    {
                        _pending.Remove(id.Value, out waiting);
                    }
                    waiting?.TrySetResult(message);
                }
            }
        }
        catch (Exception error)
        {
            Failed?.Invoke(error.Message);
        }
    }

    private void Dispatch(string name, JsonNode message)
    {
        switch (name)
        {
            case "ready":
                _ready.TrySetResult(true);
                break;
            case "status":
                StatusChanged?.Invoke(message["text"]?.GetValue<string>() ?? string.Empty);
                break;
            case "progress":
                ProgressChanged?.Invoke(message["done"]?.GetValue<int>() ?? 0,
                                        message["total"]?.GetValue<int>() ?? 1);
                break;
            case "frame":
                FrameReady?.Invoke(message["path"]?.GetValue<string>() ?? string.Empty,
                                   message["width"]?.GetValue<int>() ?? 0,
                                   message["height"]?.GetValue<int>() ?? 0);
                break;
            case "mirror_lost":
                MirrorLost?.Invoke(message["error"]?.GetValue<string>() ?? string.Empty);
                break;
        }
    }

    /// <summary>Send one request and wait for its reply.</summary>
    /// <exception cref="EngineException">The engine reported a problem.</exception>
    public async Task<JsonNode> CallAsync(string operation, object? arguments = null)
    {
        var id = Interlocked.Increment(ref _nextId);
        var request = new JsonObject { ["id"] = id, ["op"] = operation };
        if (arguments is not null)
        {
            foreach (var field in JsonSerializer.SerializeToNode(arguments)!.AsObject())
            {
                request[field.Key] = field.Value?.DeepClone();
            }
        }

        var waiting = new TaskCompletionSource<JsonNode>(TaskCreationOptions.RunContinuationsAsynchronously);
        lock (_pending)
        {
            _pending[id] = waiting;
        }

        await _writeLock.WaitAsync().ConfigureAwait(false);
        try
        {
            await _process.StandardInput.WriteLineAsync(request.ToJsonString()).ConfigureAwait(false);
            await _process.StandardInput.FlushAsync().ConfigureAwait(false);
        }
        finally
        {
            _writeLock.Release();
        }

        var reply = await waiting.Task.ConfigureAwait(false);
        if (reply["ok"]?.GetValue<bool>() != true)
        {
            throw new EngineException(reply["error"]?.GetValue<string>() ?? "the engine refused");
        }
        return reply["result"] ?? new JsonObject();
    }

    /// <summary>Send a request without waiting - for stop, which cannot queue.</summary>
    public async Task PostAsync(string operation)
    {
        await _writeLock.WaitAsync().ConfigureAwait(false);
        try
        {
            await _process.StandardInput.WriteLineAsync($"{{\"op\": \"{operation}\"}}").ConfigureAwait(false);
            await _process.StandardInput.FlushAsync().ConfigureAwait(false);
        }
        finally
        {
            _writeLock.Release();
        }
    }

    public async ValueTask DisposeAsync()
    {
        try
        {
            await PostAsync("quit").ConfigureAwait(false);
            if (!_process.WaitForExit(3000))
            {
                _process.Kill(entireProcessTree: true);
            }
        }
        catch (Exception)
        {
            // Shutting down; a dead engine is the desired outcome either way.
        }
        _process.Dispose();
    }
}

public sealed class EngineException : Exception
{
    public EngineException(string message) : base(message)
    {
    }
}

File: macos/Sources/MThreadDraw/Engine.swift
import Foundation

/// The MThread Draw engine, running as a child process and spoken to over a pipe.
///
/// Everything interesting - tracing an image, finding a device, injecting
/// touches - lives in the Python engine, which the Windows front end uses too.
/// This class is the whole of the Mac side's knowledge of it: write a JSON line,
/// wait for the reply carrying the same id, and publish anything that arrives
/// unasked.
///
/// Porting the engine to Swift would mean a second implementation of the same
/// subtleties, and they would drift apart by the second release.
final class Engine {

    enum EngineError: LocalizedError {
        case notStarted(String)
        case refused(String)

        var errorDescription: String? {
            switch self {
            case .notStarted(let detail): return "The engine did not start.\n\(detail)"
            case .refused(let detail): return detail
            }
        }
    }

    private let process = Process()
    private let input = Pipe()
    private let output = Pipe()
    private var nextId = 0
    private var pending: [Int: CheckedContinuation<[String: Any], Error>] = [:]
    private let lock = NSLock()
    private var buffer = Data()

    /// Called for events - anything the engine sends without being asked.
    var onStatus: ((String) -> Void)?
    var onProgress: ((Int, Int) -> Void)?
    var onFrame: ((String, Int, Int) -> Void)?
    var onMirrorLost: ((String) -> Void)?

    private var readyContinuation: CheckedContinuation<Void, Error>?

    /// Where the engine is, in a release and in a source checkout.
    static func locate() -> (URL, [String]) {
        let bundle = Bundle.main.bundleURL

        // A release ships the engine inside the bundle, beside the binary.
        let packaged = bundle
            .appendingPathComponent("Contents/Resources/engine/mthread-draw-engine")
        if FileManager.default.isExecutableFile(atPath: packaged.path) {
            return (packaged, [])
        }

        // A source checkout runs it from the repository's virtual environment,
        // and failing that from whatever python is on PATH.
        if let repository = findRepository() {
            let venv = repository.appendingPathComponent("venv/bin/python")
            let python = FileManager.default.isExecutableFile(atPath: venv.path)
                ? venv
                : URL(fileURLWithPath: "/usr/bin/env")
            let arguments = python.lastPathComponent == "env"
                ? ["python3", "-m", "mthread_draw.server"]
                : ["-m", "mthread_draw.server"]
            return (python, arguments)
        }
        return (URL(fileURLWithPath: "/usr/bin/env"),
                ["python3", "-m", "mthread_draw.server"])
    }

    /// Walk up from the binary looking for the engine's own source.
    ///
    /// Counting "../.." from the build directory works until somebody changes
    /// the configuration, at which point it silently points at the wrong folder.
    /// A marker file does not have that problem.
    static func findRepository() -> URL? {
        var directory = Bundle.main.bundleURL
        for _ in 0..<8 {
            let marker = directory.appendingPathComponent("mthread_draw/server.py")
            if FileManager.default.fileExists(atPath: marker.path) {
                return directory
            }
            directory = directory.deletingLastPathComponent()
        }
        return nil
    }

    func start() async throws {
        let (executable, arguments) = Engine.locate()
        process.executableURL = executable
        process.arguments = arguments
        process.standardInput = input
        process.standardOutput = output
        process.standardError = Pipe()
        if let repository = Engine.findRepository() {
            process.currentDirectoryURL = repository
        }

        output.fileHandleForReading.readabilityHandler = { [weak self] handle in
            self?.absorb(handle.availableData)
        }

        try process.run()

        // The engine announces itself; if that never arrives, nothing else will.
        try await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask { [weak self] in
                try await withCheckedThrowingContinuation { continuation in
                    self?.readyContinuation = continuation
                }
            }
            group.addTask {
                try await Task.sleep(for: .seconds(30))
                throw EngineError.notStarted("It produced no output in thirty seconds.")
            }
            try await group.next()
            group.cancelAll()
        }
    }

    // MARK: - reading

    private func absorb(_ data: Data) {
        guard !data.isEmpty else { return }
        buffer.append(data)
        while let newline = buffer.firstIndex(of: 0x0A) {
            let line = buffer[buffer.startIndex..<newline]
            buffer = buffer[buffer.index(after: newline)...]
            handle(line: Data(line))
        }
    }

    private func handle(line: Data) {
        guard !line.isEmpty,
              let message = try? JSONSerialization.jsonObject(with: line) as? [String: Any]
        else { return }

        if let name = message["event"] as? String {
            dispatch(event: name, message: message)
            return
        }

        guard let id = message["id"] as? Int else { return }
        lock.lock()
        let waiting = pending.removeValue(forKey: id)
        lock.unlock()

        guard let waiting else { return }
        if message["ok"] as? Bool == true {
            waiting.resume(returning: message["result"] as? [String: Any] ?? [:])
        } else {
            waiting.resume(throwing: EngineError.refused(
                message["error"] as? String ?? "the engine refused"))
        }
    }

    private func dispatch(event name: String, message: [String: Any]) {
        switch name {
        case "ready":
            let continuation = readyContinuation
            readyContinuation = nil
            continuation?.resume()
        case "status":
            onStatus?(message["text"] as? String ?? "")
        case "progress":
            onProgress?(message["done"] as? Int ?? 0, message["total"] as? Int ?? 1)
        case "frame":
            onFrame?(message["path"] as? String ?? "",
                     message["width"] as? Int ?? 0,
                     message["height"] as? Int ?? 0)
        case "mirror_lost":
            onMirrorLost?(message["error"] as? String ?? "")
        default:
            break
        }
    }

    // MARK: - writing

    /// Send one request and wait for its reply.
    @discardableResult
    func call(_ operation: String, _ arguments: [String: Any] = [:]) async throws -> [String: Any] {
        lock.lock()
        nextId += 1
        let id = nextId
        lock.unlock()

        var request: [String: Any] = arguments
        request["id"] = id
        request["op"] = operation
        let line = try JSONSerialization.data(withJSONObject: request) + Data([0x0A])

        return try await withCheckedThrowingContinuation { continuation in
            lock.lock()
            pending[id] = continuation
            lock.unlock()
            input.fileHandleForWriting.write(line)
        }
    }

    /// Send a request without waiting - for stop, which cannot queue.
    func post(_ operation: String) {
        let line = "{\"op\": \"\(operation)\"}\n"
        input.fileHandleForWriting.write(Data(line.utf8))
    }

    func shutDown() {
        post("quit")
        // A dead engine is the desired outcome either way.
        DispatchQueue.global().asyncAfter(deadline: .now() + 3) { [process] in
            if process.isRunning { process.terminate() }
        }
    }
}
